[feat] Improve the container platform API exposed to tests (#1775)
eth-cscs/reframe: docs/tutorial_advanced.rst

First, we need to enable the container platform support in ReFrame's configuration and, specifically, at the partition configuration level:

.. literalinclude:: ../tutorials/config/settings.py
   :lines: 38-62
   :emphasize-lines: 15-24

For each partition, users can define a list of container platforms supported using the :js:attr:`container_platforms` `configuration parameter <config_reference.html#.systems[].partitions[].container_platforms>`__.
In this case, we define the `Sarus <https://github.com/eth-cscs/sarus>`__ platform for which we set the :js:attr:`modules` parameter in order to instruct ReFrame to load the ``sarus`` module, whenever it needs to run with this container platform.

> Furthermore, we define the `Singularity <https://sylabs.io>`__ platform, for which the ``singularity`` module needs to be loaded.

Reviewer: ```suggestion
Similarly, we add an entry for the `Singularity <https://sylabs.io>`__ platform.
```

> The following parameterized test, will create two test cases, one for each of the supported contaiter platforms:

Reviewer: ```suggestion
The following parameterized test, will create two tests, one for each of the supported container platforms:
```

.. literalinclude:: ../tutorials/advanced/containers/container_test.py
   :lines: 6-
   :emphasize-lines: 11-16

A container-based test can be written as :class:`RunOnlyRegressionTest <reframe.core.pipeline.RunOnlyRegressionTest>` that sets the :attr:`container_platform <reframe.core.pipeline.RegressionTest.container_platform>` attribute.
This attribute accepts a string that corresponds to the name of the container platform that will be used to run the container for this test.
If such a platform is not `configured <config_reference.html#container-platform-configuration>`__ for the current system, the test will fail.

As soon as the container platform to be used is defined, you need to specify the container image to use by setting the :attr:`image <reframe.core.containers.ContainerPlatform.image>`.

> In the ``Singularity`` test case, we add the ``docker://`` prefix to the image name, in order to instruct ``Singularity`` to pull the image from `DockerHub <https://hub.docker.com/>`__.

Reviewer: ```suggestion
In the ``Singularity`` test variant, we add the ``docker://`` prefix to the image name, in order to instruct ``Singularity`` to pull the image from `DockerHub <https://hub.docker.com/>`__.
```

Test cases have a specific definition in ReFrame, so avoid the term here. A parameterized test generates *tests*, not *test cases*.

In this case, the test will be using `Singularity <https://sylabs.io>`__ as a container platform.
If such a platform is not `configured <config_reference.html#container-platform-configuration>`__ for the current system, the test will fail.

As soon as the container platform to be used is defined, you need to specify the container image to use by setting the :attr:`image <reframe.core.containers.ContainerPlatform.image>`.

> The default command that the container runs can be overwritten by setting the :attr:`command <reframe.core.containers.ContainerPlatform.command>` of the container platform.

Reviewer: ```suggestion
The default command that the container runs can be overwritten by setting the :attr:`command <reframe.core.containers.ContainerPlatform.command>` attribute of the container platform.
```

It is important to note that the :attr:`executable <reframe.core.pipeline.RegressionTest.executable>` and :attr:`executable_opts <reframe.core.pipeline.RegressionTest.executable_opts>` attributes of the actual test are ignored in case of container-based tests.

ReFrame will run the container as follows:

.. code-block:: console

    singularity exec -B"/path/to/test/stagedir:/rfm_workdir" --pwd="/rfm_workdir" docker://ubuntu:18.04 bash -c 'pwd; cat /etc/os-release > release.txt'

By default ReFrame will mount the stage directory of the test under ``/rfm_workdir`` inside the container.
Once the commands are executed, the container is stopped and ReFrame goes on with the sanity and performance checks.
Besides the stage directory, additional mount points can be specified through the :attr:`mount_points <reframe.core.pipeline.RegressionTest.container_platform.mount_points>` attribute:

.. code-block:: python

    self.container_platform.mount_points = [('/path/to/host/dir1', '/path/to/container/mount_point1'),
                                            ('/path/to/host/dir2', '/path/to/container/mount_point2')]

> The container filesystem is ephemeral, therefore, ReFrame mounts the stage directory under ``/rfm_workdir`` inside the container where the user can copy directories/files as needed.

Reviewer: ```suggestion
The container filesystem is ephemeral, therefore, ReFrame mounts the stage directory under ``/rfm_workdir`` inside the container where the user can copy artifacts as needed.
```

Same for the following occurences.

The aforementioned directories/files will then be available inside the stage directory after the container execution finishes.
This is very useful if the above directories/files are going to be used for the sanity/performance checks.

> If this is the case, the user should overwrite the default command executed by the container, using the :attr:`command <reframe.core.containers.ContainerPlatform.command>` to include the appropriate copy commands.

Reviewer: ```suggestion
If the copy does not happen inside the container, the user should overwrite the default command executed by the container, using the :attr:`command <reframe.core.containers.ContainerPlatform.command>` to include the appropriate copy commands.
```

> In this test, we have changed the directory that the container is going to run the commands using the `--pwd=/rfm_workdir` through the :attr:`options <reframe.core.containers.ContainerPlatform.options>`.

Reviewer: I would prefer if you add the `cp` as you just proposing above. It can be confusing.

Author: Do you mean to add also the command I use to copy the artifacts?

Reviewer: I am suggesting to not use the `--pwd`. Let the container execute on its ephemeral filesystem and then copy the  `releases.txt` to the `/rfm_workdir`, as you are exactly suggesting right above.

                                            ('/path/to/host/dir2', '/path/to/container/mount_point2')]

The container filesystem is ephemeral, therefore, ReFrame mounts the stage directory under ``/rfm_workdir`` inside the container where the user can copy artifacts as needed.

> The aforementioned artifacts will then be available inside the stage directory after the container execution finishes.

Reviewer: ```suggestion
These artifacts will therefore be available inside the stage directory after the container execution finishes.
```

> This is very useful if the above artifacts are going to be used for the sanity/performance checks.

Reviewer: ```suggestion
This is very useful if the artifacts are needed for the sanity or the performance checks.
```

> If the copy is not performed by the default container commands, the user should overwrite the default command executed by the container, using the :attr:`command <reframe.core.containers.ContainerPlatform.command>` to include the appropriate copy commands.

Reviewer: ```suggestion
If the copy is not performed by the default container command, the user can override this command by setting the :attr:`command <reframe.core.containers.ContainerPlatform.command>` attribute such as to include the appropriate copy commands.
```

In the current test, the output of the ``cat /etc/os-release`` is made available outside the container since we have used the command:

.. code-block:: bash

>     bash -c 'cat /etc/os-release | tee /rfm_workdir/release.txt'

Reviewer: You should mention perhaps that you also print in stdout to demonstrate also how the stdout of the container can be used in sanity and performance checking as usual.
